File: src/formatters/builders/amp.ts
import { getAmpExtra } from '../../providers/extras';
import type { ProviderQuota } from '../../providers/types';
import { BOX } from '../../theme';
import { barSegments, type ColorToken, colorForDisplay, indicatorSegments, type Line } from '../segments';
import { etaLabel, formatEta, formatPercent, formatResetTime, toDisplay } from '../shared';
import { buildFooterLine, labelLine, raw, vLine } from './shared';
import type { BuildOptions } from './types';

// ---------------------------------------------------------------------------
// Per-surface `labelColor` (verified against the old builders):
//   Terminal: 'magenta'  — old `label()` helper hardcoded ANSI.magenta for the ◆
//   Waybar:   'magenta'  — old label used PROVIDER_HEX.amp (ONE_DARK.magenta)
//   TUI:      'blue'     — old `label()` helper hardcoded oneDark.blue for the ◆
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Internal line helpers
// ---------------------------------------------------------------------------

/**
 * Free Tier bar line shared between sublines and inline layouts.
 * Inline layout appends the ETA directly; sublines layout omits it.
 */
function freeTierBarLine(disp: number | null, mode: BuildOptions['mode'], etaSegments: Line): Line {
  return [
    { text: BOX.v, color: 'magenta' },
    raw('  '),
    ...indicatorSegments(disp, mode),
    raw(' '),
    ...barSegments(disp, mode),
    raw(' '),
    { text: formatPercent(disp).padStart(4), color: colorForDisplay(disp, mode) },
    ...etaSegments,
  ];
}

// ---------------------------------------------------------------------------
// Main builder
// ---------------------------------------------------------------------------

/**
 * Pure builder for the Amp provider card.
 *
 * Emits Line[] with no I/O, no settings reads, no markup.
 * Surface-specific differences are driven by `options`:
 *
 * ampFreeTierLayout:
 *   'sublines' — Terminal: bar line (no ETA) + tree-connector sub-lines (├─/└─)
 *                for dollar info and ETA.
 *   'inline'   — Waybar:  bar line with inline ETA + a dot-○ line for dollar info.
 *   'generic'  — TUI:     simplified generic model loop (no Free Tier special
 *                handling; iterates all models with name+bar+pct).
 *
 * labelColor (per surface):
 *   Terminal → 'magenta', Waybar → 'magenta', TUI → 'blue'
 */
export function buildAmp(p: ProviderQuota, options: BuildOptions): Line[] {
  const { mode, headerTitle, headerWidth, labelColor, footer, ampFreeTierLayout = 'inline', accountInHeader } = options;
  const headerFill = Math.max(1, headerWidth - headerTitle.length);
  const lines: Line[] = [];

  const m: Record<string, string> = getAmpExtra(p)?.meta ?? {};

  // Header
  lines.push([
    { text: BOX.tl + BOX.h, color: 'magenta' },
    raw(' '),
    { text: headerTitle, color: 'magenta', bold: true },
    raw(' '),
    { text: BOX.h.repeat(headerFill), color: 'magenta' },
  ]);
  lines.push(vLine('magenta'));

  if (p.error) {
    lines.push([{ text: BOX.v, color: 'magenta' }, raw('  '), { text: `⚠️ ${p.error}`, color: 'red' }]);
  } else if (ampFreeTierLayout === 'generic') {
    // -----------------------------------------------------------------------
    // TUI layout: generic model loop (no Free Tier special handling)
    // -----------------------------------------------------------------------
    if (!p.models || Object.keys(p.models).length === 0) {
      lines.push([{ text: BOX.v, color: 'magenta' }, raw('  '), { text: 'No usage data', color: 'muted' }]);
    } else {
      const entries = Object.entries(p.models);
      const maxLen = Math.max(...entries.map(([name]) => name.length), 20);

      lines.push(labelLine('Usage', labelColor, 'magenta'));
      for (const [name, window] of entries) {
        const disp = toDisplay(window.remaining, mode);
        lines.push([
          { text: BOX.v, color: 'magenta' },
          raw('  '),
          ...indicatorSegments(disp, mode),
          raw(' '),
          { text: name.padEnd(maxLen), color: 'textBright' },
          raw(' '),
          ...barSegments(disp, mode),
          raw(' '),
          { text: formatPercent(disp).padStart(4), color: colorForDisplay(disp, mode) },
        ]);
      }
    }
  } else {
    // -----------------------------------------------------------------------
    // Terminal ('sublines') and Waybar ('inline') layouts
    // -----------------------------------------------------------------------
    const free = p.models?.['Free Tier'];

    if (free) {
      const rem = free.remaining;
      const disp = toDisplay(rem, mode);

      lines.push(labelLine('Free Tier', labelColor, 'magenta'));

      if (ampFreeTierLayout === 'sublines') {
        // Terminal: bar line without ETA; sub-lines with tree connectors
        lines.push(freeTierBarLine(disp, mode, []));

        // Build sub-details
        const subs: Line[] = [];

        // Dollar info sub-line: replenishRate  ( freeRemaining / freeTotal )  bonus
        const dollarParts: Line = [];
        if (m.replenishRate) {
          dollarParts.push({ text: m.replenishRate, color: 'cyan' });
        }
        const dollars = [m.freeRemaining, m.freeTotal].filter(Boolean).join(' / ');
        if (dollars) {
          if (dollarParts.length > 0) dollarParts.push(raw('  '));
          dollarParts.push({ text: `( ${dollars} )`, color: 'text' });
        }
        if (m.bonus) {
          if (dollarParts.length > 0) dollarParts.push(raw('  '));
          dollarParts.push({ text: m.bonus, color: 'cyan' });
        }
        if (dollarParts.length > 0) subs.push(dollarParts);

        // ETA sub-line (only when resetsAt is present and not full)
        if (free.resetsAt && rem !== 100) {
          const etaText = `${etaLabel(mode)} ${formatEta(free.resetsAt, rem)}  ${formatResetTime(free.resetsAt, rem)}`;
          subs.push([{ text: etaText, color: 'cyan' }]);
        }

        for (let i = 0; i < subs.length; i++) {
          const conn = i === subs.length - 1 ? '└─' : '├─';
          lines.push([
            { text: BOX.v, color: 'magenta' },
            raw('  '),
            { text: conn, color: 'comment' },
            raw(' '),
            ...subs[i],
          ]);
        }
      } else {
        // Waybar inline: bar line with ETA appended; dollar info on ○ line
        const etaSegs: Line =
          free.resetsAt && rem !== 100
            ? [
                raw('  '),
                {
                  text: `→ ${etaLabel(mode)} ${formatEta(free.resetsAt, rem)} ${formatResetTime(free.resetsAt, rem)}`,
                  color: 'cyan',
                },
              ]
            : [];

        lines.push(freeTierBarLine(disp, mode, etaSegs));

        // Dollar info on ○ line
        const infoParts: Line = [];
        if (m.replenishRate) {
          infoParts.push({ text: m.replenishRate, color: 'cyan' });
        }
        const dollars = [m.freeRemaining, m.freeTotal].filter(Boolean).join(' / ');
        if (dollars) {
          if (infoParts.length > 0) infoParts.push({ text: '  |  ', color: 'comment' });
          infoParts.push({ text: dollars, color: 'text' });
        }
        if (m.bonus) {
          if (infoParts.length > 0) infoParts.push({ text: '  |  ', color: 'comment' });
          infoParts.push({ text: m.bonus, color: 'cyan' });
        }
        if (infoParts.length > 0) {
          lines.push([
            { text: BOX.v, color: 'magenta' },
            raw('  '),
            { text: BOX.dotO, color: 'comment' },
            raw(' '),
            ...infoParts,
          ]);
        }
      }
    }

    // Credits section (terminal + waybar; not in TUI generic layout)
    const credits = p.models?.Credits;
    if (credits) {
      lines.push(vLine('magenta'));
      const balance = m.creditsBalance ?? '$0';
      const creditColor: ColorToken = credits.remaining > 0 ? 'green' : 'comment';
      lines.push(labelLine('Credits', labelColor, 'magenta'));
      const creditDisp = toDisplay(credits.remaining, mode);
      lines.push([
        { text: BOX.v, color: 'magenta' },
        raw('  '),
        ...indicatorSegments(creditDisp, mode),
        raw(' '),
        {
          text: ampFreeTierLayout === 'inline' ? `${balance} remaining` : balance,
          color: creditColor,
        },
      ]);
    }

    // Fallback for unknown models (when neither Free Tier nor Credits)
    if (!free && !credits && p.models && Object.keys(p.models).length > 0) {
      const entries = Object.entries(p.models);
      const maxLen = Math.max(...entries.map(([name]) => name.length), 20);
      lines.push(labelLine('Usage', labelColor, 'magenta'));
      for (const [name, window] of entries) {
        const disp = toDisplay(window.remaining, mode);
        lines.push([
          { text: BOX.v, color: 'magenta' },
          raw('  '),
          ...indicatorSegments(disp, mode),
          raw(' '),
          { text: name.padEnd(maxLen), color: 'textBright' },
          raw(' '),
          ...barSegments(disp, mode),
          raw(' '),
          { text: formatPercent(disp).padStart(4), color: colorForDisplay(disp, mode) },
        ]);
      }
    }
  }

  // Account line — omitted when the surface already shows the account in the header
  if (p.account && !accountInHeader) {
    lines.push(vLine('magenta'));
    lines.push([{ text: BOX.v, color: 'magenta' }, raw('  '), { text: `Account: ${p.account}`, color: 'comment' }]);
  }

  lines.push(vLine('magenta'));
  lines.push(buildFooterLine(footer, 'magenta'));

  return lines;
}

File: rust/src/formatters/builders/amp.rs
//! Builder do card Amp. Port fiel de `src/formatters/builders/amp.ts`.
//! Cor de marca = Magenta. Três layouts de Free Tier: Generic/Sublines/Inline.

use std::collections::BTreeMap;

use crate::formatters::clock::Clock;
use crate::formatters::segments::{
    bar_segments, color_for_display, indicator_segments, Line, Segment,
};
use crate::formatters::shared::{
    eta_label, format_eta, format_percent, format_reset_time, to_display,
};
use crate::providers::extras::get_amp_extra;
use crate::providers::types::ProviderQuota;
use crate::settings::DisplayMode;
use crate::theme::{box_chars, ColorToken};

use super::shared::{build_footer_line, header_line, label_line, vline, AmpLayout, BuildOptions};

/// Valor de meta com semântica truthy do JS: Some só se presente E não-vazio.
fn meta_get<'a>(m: &'a BTreeMap<String, String>, k: &str) -> Option<&'a str> {
    m.get(k).map(|s| s.as_str()).filter(|s| !s.is_empty())
}

/// Junta freeRemaining/freeTotal (truthy) com " / ".
fn dollars(m: &BTreeMap<String, String>) -> String {
    [meta_get(m, "freeRemaining"), meta_get(m, "freeTotal")]
        .into_iter()
        .flatten()
        .collect::<Vec<_>>()
        .join(" / ")
}

/// Linha da barra do Free Tier (compartilhada por sublines/inline). `eta_segments`
/// é anexado ao fim (inline anexa ETA; sublines passa vazio).
fn free_tier_bar_line(disp: Option<f64>, mode: DisplayMode, eta_segments: Line) -> Line {
    let mut line: Line = vec![
        Segment::new(box_chars::V, ColorToken::Magenta),
        Segment::raw_text("  "),
    ];
    line.extend(indicator_segments(disp, mode));
    line.push(Segment::raw_text(" "));
    line.extend(bar_segments(disp, mode));
    line.push(Segment::raw_text(" "));
    line.push(Segment::new(
        format!("{:>4}", format_percent(disp)),
        color_for_display(disp, mode),
    ));
    line.extend(eta_segments);
    line
}

/// Linha genérica de model (indicador + nome + barra + pct), sem ETA.
fn generic_model_line(name: &str, max_len: usize, disp: Option<f64>, mode: DisplayMode) -> Line {
    let mut line: Line = vec![
        Segment::new(box_chars::V, ColorToken::Magenta),
        Segment::raw_text("  "),
    ];
    line.extend(indicator_segments(disp, mode));
    line.push(Segment::raw_text(" "));
    line.push(Segment::new(
        format!("{name:<max_len$}"),
        ColorToken::TextBright,
    ));
    line.push(Segment::raw_text(" "));
    line.extend(bar_segments(disp, mode));
    line.push(Segment::raw_text(" "));
    line.push(Segment::new(
        format!("{:>4}", format_percent(disp)),
        color_for_display(disp, mode),
    ));
    line
}

pub fn build_amp(clock: &Clock, p: &ProviderQuota, options: &BuildOptions) -> Vec<Line> {
    let mode = options.mode;
    let layout = options.amp_free_tier_layout;
    let empty_meta = BTreeMap::new();
    let m: &BTreeMap<String, String> = get_amp_extra(p)
        .and_then(|e| e.meta.as_ref())
        .unwrap_or(&empty_meta);

    let mut lines: Vec<Line> = Vec::new();

    lines.push(header_line(
        &options.header_title,
        options.header_width,
        ColorToken::Magenta,
    ));
    lines.push(vline(ColorToken::Magenta));

    if let Some(err) = p.error.as_deref() {
        lines.push(vec![
            Segment::new(box_chars::V, ColorToken::Magenta),
            Segment::raw_text("  "),
            Segment::new(format!("⚠️ {err}"), ColorToken::Red),
        ]);
    } else if layout == AmpLayout::Generic {
        // TUI: loop genérico, sem special-casing de Free Tier/Credits.
        match p.models.as_ref().filter(|mm| !mm.is_empty()) {
            None => lines.push(vec![
                Segment::new(box_chars::V, ColorToken::Magenta),
                Segment::raw_text("  "),
                Segment::new("No usage data", ColorToken::Muted),
            ]),
            Some(models) => {
                let max_len = models
                    .keys()
                    .map(|n| n.chars().count())
                    .max()
                    .unwrap_or(0)
                    .max(20);
                lines.push(label_line(
                    "Usage",
                    options.label_color,
                    ColorToken::Magenta,
                ));
                for (name, window) in models {
                    let disp = to_display(Some(window.remaining), mode);
                    lines.push(generic_model_line(name, max_len, disp, mode));
                }
            }
        }
    } else {
        // Terminal (sublines) e Waybar (inline).
        let free = p.models.as_ref().and_then(|mm| mm.get("Free Tier"));

        if let Some(free) = free {
            let rem = free.remaining;
            let disp = to_display(Some(rem), mode);
            lines.push(label_line(
                "Free Tier",
                options.label_color,
                ColorToken::Magenta,
            ));

            if layout == AmpLayout::Sublines {
                lines.push(free_tier_bar_line(disp, mode, Vec::new()));

                let mut subs: Vec<Line> = Vec::new();

                // Sub-linha de dólares: replenishRate  ( freeRemaining / freeTotal )  bonus
                let mut dollar_parts: Line = Vec::new();
                if let Some(rate) = meta_get(m, "replenishRate") {
                    dollar_parts.push(Segment::new(rate.to_string(), ColorToken::Cyan));
                }
                let d = dollars(m);
                if !d.is_empty() {
                    if !dollar_parts.is_empty() {
                        dollar_parts.push(Segment::raw_text("  "));
                    }
                    dollar_parts.push(Segment::new(format!("( {d} )"), ColorToken::Text));
                }
                if let Some(bonus) = meta_get(m, "bonus") {
                    if !dollar_parts.is_empty() {
                        dollar_parts.push(Segment::raw_text("  "));
                    }
                    dollar_parts.push(Segment::new(bonus.to_string(), ColorToken::Cyan));
                }
                if !dollar_parts.is_empty() {
                    subs.push(dollar_parts);
                }

                // Sub-linha de ETA (só com resets e não cheio).
                if free.resets_at.is_some() && rem != 100.0 {
                    let eta_text = format!(
                        "{} {}  {}",
                        eta_label(mode),
                        format_eta(clock, free.resets_at.as_deref(), rem),
                        format_reset_time(clock, free.resets_at.as_deref(), rem)
                    );
                    subs.push(vec![Segment::new(eta_text, ColorToken::Cyan)]);
                }

                let last = subs.len().saturating_sub(1);
                for (i, sub) in subs.into_iter().enumerate() {
                    let conn = if i == last { "└─" } else { "├─" };
                    let mut line: Line = vec![
                        Segment::new(box_chars::V, ColorToken::Magenta),
                        Segment::raw_text("  "),
                        Segment::new(conn, ColorToken::Comment),
                        Segment::raw_text(" "),
                    ];
                    line.extend(sub);
                    lines.push(line);
                }
            } else {
                // Inline (Waybar): barra com ETA anexado; dólares na linha ○.
                let eta_segs: Line = if free.resets_at.is_some() && rem != 100.0 {
                    vec![
                        Segment::raw_text("  "),
                        Segment::new(
                            format!(
                                "→ {} {} {}",
                                eta_label(mode),
                                format_eta(clock, free.resets_at.as_deref(), rem),
                                format_reset_time(clock, free.resets_at.as_deref(), rem)
                            ),
                            ColorToken::Cyan,
                        ),
                    ]
                } else {
                    Vec::new()
                };
                lines.push(free_tier_bar_line(disp, mode, eta_segs));

                let mut info_parts: Line = Vec::new();
                if let Some(rate) = meta_get(m, "replenishRate") {
                    info_parts.push(Segment::new(rate.to_string(), ColorToken::Cyan));
                }
                let d = dollars(m);
                if !d.is_empty() {
                    if !info_parts.is_empty() {
                        info_parts.push(Segment::new("  |  ", ColorToken::Comment));
                    }
                    info_parts.push(Segment::new(d, ColorToken::Text));
                }
                if let Some(bonus) = meta_get(m, "bonus") {
                    if !info_parts.is_empty() {
                        info_parts.push(Segment::new("  |  ", ColorToken::Comment));
                    }
                    info_parts.push(Segment::new(bonus.to_string(), ColorToken::Cyan));
                }
                if !info_parts.is_empty() {
                    let mut line: Line = vec![
                        Segment::new(box_chars::V, ColorToken::Magenta),
                        Segment::raw_text("  "),
                        Segment::new(box_chars::DOT_O, ColorToken::Comment),
                        Segment::raw_text(" "),
                    ];
                    line.extend(info_parts);
                    lines.push(line);
                }
            }
        }

        // Credits (terminal + waybar).
        let credits = p.models.as_ref().and_then(|mm| mm.get("Credits"));
        if let Some(credits) = credits {
            lines.push(vline(ColorToken::Magenta));
            let balance = m.get("creditsBalance").map(|s| s.as_str()).unwrap_or("$0");
            let credit_color = if credits.remaining > 0.0 {
                ColorToken::Green
            } else {
                ColorToken::Comment
            };
            lines.push(label_line(
                "Credits",
                options.label_color,
                ColorToken::Magenta,
            ));
            let credit_disp = to_display(Some(credits.remaining), mode);
            let balance_text = if layout == AmpLayout::Inline {
                format!("{balance} remaining")
            } else {
                balance.to_string()
            };
            let mut line: Line = vec![
                Segment::new(box_chars::V, ColorToken::Magenta),
                Segment::raw_text("  "),
            ];
            line.extend(indicator_segments(credit_disp, mode));
            line.push(Segment::raw_text(" "));
            line.push(Segment::new(balance_text, credit_color));
            lines.push(line);
        }

        // Fallback p/ models desconhecidos (nem Free Tier nem Credits).
        if free.is_none() && credits.is_none() {
            if let Some(models) = p.models.as_ref().filter(|mm| !mm.is_empty()) {
                let max_len = models
                    .keys()
                    .map(|n| n.chars().count())
                    .max()
                    .unwrap_or(0)
                    .max(20);
                lines.push(label_line(
                    "Usage",
                    options.label_color,
                    ColorToken::Magenta,
                ));
                for (name, window) in models {
                    let disp = to_display(Some(window.remaining), mode);
                    lines.push(generic_model_line(name, max_len, disp, mode));
                }
            }
        }
    }

    // Account line — omitida quando a superfície já mostra a conta no header.
    if let Some(account) = p.account.as_deref().filter(|s| !s.is_empty()) {
        if !options.account_in_header {
            lines.push(vline(ColorToken::Magenta));
            lines.push(vec![
                Segment::new(box_chars::V, ColorToken::Magenta),
                Segment::raw_text("  "),
                Segment::new(format!("Account: {account}"), ColorToken::Comment),
            ]);
        }
    }

    lines.push(vline(ColorToken::Magenta));
    lines.push(build_footer_line(
        clock,
        options.footer_fetched_at.as_deref(),
        ColorToken::Magenta,
    ));

    lines
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::formatters::builders::shared::{AmpLayout, BuildOptions};
    use crate::formatters::clock::Clock;
    use crate::formatters::render_pango::render_pango;
    use crate::providers::types::{AmpQuotaExtra, ProviderExtra, ProviderQuota, QuotaWindow};
    use crate::settings::DisplayMode;
    use crate::theme::ColorToken;
    use std::collections::BTreeMap;
    use time::macros::datetime;

    fn clk() -> Clock {
        Clock {
            now: datetime!(2026-06-19 12:00:00 UTC),
            local_offset: time::UtcOffset::UTC,
        }
    }

    fn opts(layout: AmpLayout) -> BuildOptions {
        BuildOptions {
            mode: DisplayMode::Remaining,
            header_title: "Amp".into(),
            header_width: 52,
            label_color: ColorToken::Magenta,
            footer_fetched_at: None,
            plan_label: None,
            amp_free_tier_layout: layout,
            account_in_header: false,
        }
    }

    fn win(r: f64) -> QuotaWindow {
        QuotaWindow {
            remaining: r,
            resets_at: Some("2026-06-19T14:00:00Z".into()),
            window_minutes: None,
            used: None,
        }
    }

    fn amp_with_free_and_credits() -> ProviderQuota {
        let mut models = BTreeMap::new();
        models.insert("Free Tier".to_string(), win(30.0));
        models.insert("Credits".to_string(), win(75.0));
        let mut meta = BTreeMap::new();
        meta.insert("freeRemaining".to_string(), "$1.50".to_string());
        meta.insert("freeTotal".to_string(), "$5.00".to_string());
        meta.insert("replenishRate".to_string(), "+$0.10/h".to_string());
        meta.insert("creditsBalance".to_string(), "$12".to_string());
        ProviderQuota {
            provider: "amp".into(),
            display_name: "Amp".into(),
            available: true,
            account: Some("me@x.com".into()),
            plan: None,
            plan_type: None,
            primary: Some(win(30.0)),
            secondary: None,
            models: Some(models),
            extra: Some(ProviderExtra::Amp(AmpQuotaExtra { meta: Some(meta) })),
            error: None,
        }
    }

    #[test]
    fn inline_layout_has_free_tier_and_credits_and_account() {
        let out = render_pango(&build_amp(
            &clk(),
            &amp_with_free_and_credits(),
            &opts(AmpLayout::Inline),
        ));
        assert!(out.contains("Free Tier"));
        assert!(out.contains("Credits"));
        assert!(out.contains("$12 remaining")); // inline anexa " remaining"
        assert!(out.contains("Account: me@x.com"));
        // ○ line de dólares (inline usa separador "  |  ")
        assert!(out.contains("$1.50 / $5.00"));
    }

    #[test]
    fn sublines_layout_uses_tree_connectors() {
        let out = render_pango(&build_amp(
            &clk(),
            &amp_with_free_and_credits(),
            &opts(AmpLayout::Sublines),
        ));
        assert!(out.contains("Free Tier"));
        // sublines: dólares entre parênteses + último connector └─
        assert!(out.contains("( $1.50 / $5.00 )"));
        assert!(out.contains("└─"));
        // sublines NÃO anexa " remaining" ao balance
        assert!(out.contains("$12"));
    }

    #[test]
    fn generic_layout_loops_models() {
        // generic ignora special-casing de Free Tier/Credits e itera p.models.
        let q = amp_with_free_and_credits();
        let out = render_pango(&build_amp(&clk(), &q, &opts(AmpLayout::Generic)));
        assert!(out.contains("Usage"));
        // Account line é independente do layout → presente também no generic.
        assert!(out.contains("Account: me@x.com"));
    }

    #[test]
    fn account_omitted_when_in_header() {
        let mut o = opts(AmpLayout::Inline);
        o.account_in_header = true;
        let out = render_pango(&build_amp(&clk(), &amp_with_free_and_credits(), &o));
        assert!(!out.contains("Account:"));
    }

    #[test]
    fn error_branch() {
        let mut q = amp_with_free_and_credits();
        q.error = Some("rate limited".into());
        let out = render_pango(&build_amp(&clk(), &q, &opts(AmpLayout::Inline)));
        assert!(out.contains("⚠️ rate limited"));
        assert!(!out.contains("Free Tier"));
    }
}
